Services/Mentor/OJS.Services.Mentor.Business/Implementations/MentorBusinessService.cs

```diff
@@ -0,0 +1,601 @@
+            .ToList();
+
+        /*
+         *  There are 2 cases when it comes to document retrieval:
+         *  1. The problem has its own problem resource ( Word file ) ( e.g. online / onsite exam ).
+         *  2. All the problems' descriptions are in a single Word file ( e.g. Lab / Exercise ).
+         */
+        var problemsDescription = wordFiles.FirstOrDefault(pr => pr.ProblemId == model.ProblemId) ?? wordFiles.FirstOrDefault();
+
+        if (problemsDescription is null)
+        {
+            throw new BusinessServiceException(DocumentNotFoundOrEmpty);
+        }
+
+        var file = problemsDescription.File ??
+            (problemsDescription.Link is not null
+            ? await this.DownloadDocument(problemsDescription.Link, model.ProblemId, model.ContestId)
+            : []);
+
+        var number = GetProblemNumber(model.ProblemName);
+        var text = this.ExtractSectionFromDocument(file, model.ProblemName, number, model.ProblemId, model.ContestId);
+
+        return new ConversationMessageModel
+        {
+            Content = string.Format(
+                CultureInfo.InvariantCulture,
+                template!.Template,
+                model.ProblemName,
+                text,
+                model.ContestName,
+                model.CategoryName),
+            Role = MentorMessageRole.System,
+            // The system message should always be first ( in ascending order )
+            SequenceNumber = int.MinValue,
+            ProblemId = model.ProblemId,
+        };
+    }
+
+    private async Task<byte[]> DownloadDocument(string link, int problemId, int contestId)
+    {
+        var fileBytes = link.Contains("/svn", StringComparison.OrdinalIgnoreCase)
+            ? await this.DownloadSvnResource(link, problemId, contestId)
+            : await this.DownloadResource(link, problemId, contestId);
+
+        if (!IsExpectedFormat(fileBytes))
+        {
+            this.logger.LogInvalidDocumentFormat(problemId, contestId, link);
+            throw new BusinessServiceException(DocumentNotFoundOrEmpty);
+        }
+
+        return fileBytes;
+    }
+
+    private async Task<byte[]> DownloadResource(string url, int problemId, int contestId)
+    {
+        var client = this.httpClientFactory.CreateClient();
+        return await this.FetchResource(url, client, problemId, contestId);
+    }
+
+    private async Task<byte[]> DownloadSvnResource(string path, int problemId, int contestId)
+    {
+        var svnConfig = this.configuration.GetSectionWithValidation<SvnConfig>();
+
+        var pathParts = path.Split("svn", StringSplitOptions.RemoveEmptyEntries);
+
+        var svnUrl = $"{svnConfig.BaseUrl}{pathParts.Last()}";
+
+        var handler = new HttpClientHandler
+        {
+            Credentials = new NetworkCredential(svnConfig.Username, svnConfig.Password)
+        };
+
+        using var client = new HttpClient(handler);
```

Comment:
It's not good to create `new HttpClient` on every request, as this can lead to socket exhaustion and is also resource intensive. Much better is to create a named/typed client on StartUp with this configuration/handler and reuse it. See `LokiHttpClientName` or `SulsPlatformHttpClientService`

Reply:
Done.